holidayextras/doc-warrior
# OPS-2484-2 Process sub directories

This work makes a few changes to the `notEqual` operator to make it take into account the options in the full array. It also adds a new function to get all documents from the terms database with the parent directory, allowing us to process sub directories and build documents up dynamically.

**Changes to notEqual**
This operator previously was the reverse implementation of the Equals operator, which theoretically made sense but in practise never worked. `Equals` will go through an array of brands (for example) and would return true as soon as it found a match. 

Taking this into account for notEqual, it would loop through the array and return true as soon as it `didn't` find a match, meaning that if the first item in the array was not a match, it returned early, and would therefore not take into account anything else in the array. 

So the first thing I have done is make the `notEqual` operator build up an array of `true/false` values when comparing against each item in the array, and it will then return if the array includes `true`. 

**Changes to docs.js and the database connector file**
As Doc warrior already nicely processes all documents based on an array of directories that you pass it, all I needed to do was add the sub directories to the array that gets processed. To do this I have added a new function in the database file to get all documents with the parent directory name in their name, and add those to the array of docs to process.

This works well and only adds a few new functions to get this done. I would like to add an async waterfall to the docs.js file, however this can be done at a later stage due to the urgency of this work now, and the current way isn't the worst thing ever.

**To test this**
1. Delete everything from your local terms database. Use Sequel Pro to run the command `DELETE FROM versions` against the versions table in the terms database. 
2. Update your local terms and conditions database. Do this by doing a `git pull` on the `terms-and-conditions` repo, and then run `node build/index.js`.
3. Pull down this branch locally in Doc Warrior
4. Do an npm link from your local hapi to local doc warrior, so that you can see these changes when running hapi locally. You can do this by running this command in your local hapi terminal `npm link ../doc-warrior`
5. Fire up `local hapi` and `local Render` and get to the payment page for any of these Heathrow products `LHU3, LHU4, LHU5, LHH0, LHH1, LHH2, LHH3`. Click on the terms and conditions link on the payment page form. Under the **Before you go on holiday** heading you should see a new heading specifically for these products which says **THESE ARE THE LHR TERMS**
6. Do the same as the above but for a different product, not at Heathrow. You should not see the **THESE ARE THE LHR TERMS** heading, but instead see the normal heading about cancellations.

This shows that we are building this document up based on what product code is sent in. Also heck Heathrow products that are not in that list above to make sure that they show the **default** terms and not the Heathrow specific ones.

**Recordit**
http://recordit.co/f5hgGINmG4

![](https://media.giphy.com/media/pc7K6wz7GHt1C/giphy.gif)
